### ISSUE TITLE
Customizing icon_color not working after browser refresh

### ISSUE THREAD

**Issue by ed-holland** [Mariusthvdb/custom-ui#29]
I'm using the HomeMatic integration and was trying to show correct color icons.
There is no reaction to the customizing implemented.
![image](https://user-images.githubusercontent.com/53995991/103139112-60c06080-46d9-11eb-88e1-e6c7c44858c5.png)

The entities do not seem to have a unique id ...
![image](https://user-images.githubusercontent.com/53995991/103139121-733a9a00-46d9-11eb-87f2-f961ab380486.png)

customizing:
```
sensor.hm_low_battery:
  templates:
    icon_color: if (state == 'O.K.') return 'green'; return 'orange';    
```

Any idea why this is not working?

**Comment by Mariusthvdb** [comment]
What is the state in developer-tools/states

**Comment by ed-holland** [comment]
yeah, using (comment was more textual)
```
  customize_domain: !include domain_customize.yaml
  customize: !include_dir_merge_named customize/
  customize_glob:
# All Entities - Hide Templates & Icon Color
    "*.*":
      hide_attributes:
        - templates
        - icon_color
```
Not so sure about the order to use on the three...

Checking it now a few hours later, it seems to be fine. Caching?

I noticed that when using the chrome-refresh button, all custom-ui changes disappear.

To answer your question:
![image](https://user-images.githubusercontent.com/53995991/103143254-a2213200-4712-11eb-9ecb-1dea71ddfa78.png)

**Comment by Mariusthvdb** [comment]
```
sensor.hm_low_battery:
  templates:
    icon_color: >
      if (state == 'O.K.') return 'green';
      return 'orange';   
```

should work, on the condition custom_ui is installed correctly. Can you confirm to see that on the info page?

**Comment by ed-holland** [comment]
After adding the last-update information, it became clear that the config changes are only enabled on the next update of the entity.
For Homematic that can be quite a while depending on the entity.

Can you confirm that this is the case?

**Comment by Mariusthvdb** [comment]
Not sure what you are talking about, but no matter of state, the template should work, as long as all is installed correctly.

So, please check the post above and confirm/ show a screenshot of the info page where custom-ui should display the installed version.

**Comment by ed-holland** [comment]
The template showed up after rebooting the Redmatic PI (runs HomeMatic).
![image](https://user-images.githubusercontent.com/53995991/103155478-b27fee00-47a0-11eb-8165-3936f4dadd55.png)

Still...  starting situation
![image](https://user-images.githubusercontent.com/53995991/103155494-d6433400-47a0-11eb-8a81-f01f099be3b4.png)
Then doing a refresh in the browser
![image](https://user-images.githubusercontent.com/53995991/103155503-e3f8b980-47a0-11eb-8959-de0c4f3a0f6f.png)

**Comment by Mariusthvdb** [comment]
Let me ask a final (3d) time: has custom-ui been installed, and show the info page please.

Since there is no attribute icon_color displayed in the attributes list, I suspect that to be the issue

**Comment by ed-holland** [comment]
Sorry, of course it is installed. It is working nicely most of the time.
There is an icon_color defined in templates.

Inspect Screenshot.
![image](https://user-images.githubusercontent.com/53995991/103166628-52cb2680-4824-11eb-944f-de66a495ace0.png)

The information:
![image](https://user-images.githubusercontent.com/53995991/103166730-31b70580-4825-11eb-9139-ec521e7b5b09.png)
![image](https://user-images.githubusercontent.com/53995991/103166737-3e3b5e00-4825-11eb-8f58-4ff03cebad9a.png)
After chrome-refresh
![image](https://user-images.githubusercontent.com/53995991/103166739-485d5c80-4825-11eb-9193-882de65f77d4.png)

**Comment by Mariusthvdb** [comment]
ok, in the screenshot of the states, no icon_color attributes is displayed, which seems to indicate the template doesnt get processed. see the attribute icon_color at the bottom:

<img width="832" alt="Schermafbeelding 2020-12-27 om 11 21 35" src="https://user-images.githubusercontent.com/33354141/103168704-b873de80-4835-11eb-915d-1dd3272fbd15.png">

can you check if a non templated icon_color is displayed?
simply set a fixed icon_color for an entity.

also, is this behavior identical in each browser? and the HA app

**Comment by ed-holland** [comment]
sure.
```
sensor.hm_low_battery:
  icon_color: blue
  # templates:
  #   icon_color: if (state == 'O.K.') return 'green'; return 'orange';
```
![image](https://user-images.githubusercontent.com/53995991/103172664-51662200-4855-11eb-9a80-13f1a31fbddf.png)

![image](https://user-images.githubusercontent.com/53995991/103172676-5d51e400-4855-11eb-945b-111bd789f659.png)
After refresh
![image](https://user-images.githubusercontent.com/53995991/103172826-98084c00-4856-11eb-9a43-b9a628602b28.png)


I've additionally check with Firefox and the app and both show the colors correctly. Refresh in FireFox has the same result as in Chrome.

There are no errors shown in inspect. Also not when the chrome refresh was issued.
![image](https://user-images.githubusercontent.com/53995991/103172789-419b0d80-4856-11eb-8455-d52d6aa30929.png)

**Comment by ed-holland** [comment]
One more observation i made.
If i select another dashboard path and refresh it, going back to the X path shows the colors again.

**Comment by Mariusthvdb** [comment]
still cant reproduce this. In 3 browsers, using 3 different HA configurations, with 3 dashboards in 1 setup. Working 100%.

What bothers me in your config is the fact that the states scree shot doesnt show an icon_color attribute below the template in the attributes pane. This would indicate the attribute isn't set at all.

You could have a look before refreshing, because, as you say upon first load it does show the correct color. If it does, check the states/attributes page for the attribute icon_color please. And repeat after reload.

as a last resort (shouldn't be necessary, but one never knows) you could restart the instance, and see if that helps. You might have dine that already, but we didnt mention it before I think.

**Comment by ed-holland** [comment]
It is indeed a strange issue. Nothing major but strange. This is my setup (on a PI4)
```
Version | 2020.12.1
-- | --
Installation Type | Home Assistant Supervised
Development | false
Supervisor | true
Docker | true
Virtual Environment | false
Python Version | 3.8.6
Operating System Family | Linux
Operating System Version | 5.4.79-v7l+
CPU Architecture | armv7l
Timezone | Europe/Berlin
```
I've had to restart several times as the homemactic entities do not get customizing updates when one uses "RELOAD Customizing".

I have removed the hide_attributes and rebooted.

Developer Tools - Entity State
![image](https://user-images.githubusercontent.com/53995991/103267479-e3833d00-49b1-11eb-9270-024012678864.png)

Details of Entity (Lovelace)
![image](https://user-images.githubusercontent.com/53995991/103267550-04e42900-49b2-11eb-9b19-1976b7897224.png)

After a refresh, the details screen is still correct. Just not the dashboard.
![image](https://user-images.githubusercontent.com/53995991/103267626-33fa9a80-49b2-11eb-855d-50a67df6b317.png)

**Comment by Mariusthvdb** [comment]
if you still have:
```
     hide_attributes:
        - templates
        - icon_color
 ```

the more-info should not show the templates at all, using the latest version https://github.com/Mariusthvdb/custom-ui/releases/tag/20201218-b.
please check.

**Comment by ed-holland** [comment]
You misunderstood. `I have removed the hide_attributes and rebooted.` so that they became visible.

I'm using the latest release.

**Comment by Mariusthvdb** [comment]
Didn't misunderstood.. just checked because you hadn't stipulated ;-)

Just kidding. If anything, it at least shows that aspect is working as it should.

**Comment by ACrazyConcept** [comment]
Pretty sure I have same issue. Works with entities from e.g. tado sensor. But not for plant, which is created in yaml.

**Comment by Mariusthvdb** [comment]
suddenly this reminds me of another type of entities that cant be customized using custom_ui:  in python scripts states can be created using `hass.states.set()` .
the only way to set an icon_color on those entities is by setting it in the script itself. like:
```
    hass.states.set(entity_id, state,{
      'friendly_name': friendlyName,
      'device_class': 'problem',
      'icon_color':iconColor,
      })
```

could it be the issues at hand here are similar?

**Comment by ed-holland** [comment]
> Omt: in your topic you mention Unique IDs. Do entities with a unique id behave correctly?

Initially i thought that that was the issue. It is more related to the update of the entity. 
I will changed the title accordingly.

Example: i have a status dashboard just with entities. Nicely colored.
![image](https://user-images.githubusercontent.com/53995991/103453547-8a7e1680-4cdb-11eb-8afe-71b94fe5be18.png)

After chrome refresh:
![image](https://user-images.githubusercontent.com/53995991/103453554-9669d880-4cdb-11eb-9e68-337397c8b96a.png)

After 15 seconds or so:
![image](https://user-images.githubusercontent.com/53995991/103453565-a8e41200-4cdb-11eb-9b6c-643ac8a3473b.png)

An icon color appears. After 15s more, the Shelly GAS has color again.
It seems that some entity update triggers the customization to show again.

**Comment by dmatora** [comment]
> Initially i thought that that was the issue. It is more related to the update of the entity.
> I will changed the title accordingly.

Now it's a duplicate of #30 :)

**Comment by dmatora** [comment]
Closed #30 as duplicate of this one

**Comment by Mariusthvdb** [comment]
glad we have 1 issue less, thanks for that.

Trying to reproduce, I cant make it happen. All of my dashboards in Chrome show fine, no matter how many times I hit refresh. This is on Mac Chrome btw.

Might be am off guess, but seeing the states of your entities, they seem rather unusual. You dont happen to have customized the states of these entities? Because that can cause strange (and potentially dangerous results). It these states are the result of regular template sensors, the should be alright, but just like to be sure.

do all entities get their customization back after a few seconds after the refresh, or are there still entities left uncustomized?

**Comment by dmatora** [comment]
Still left
Cache clear didn't help
AnyDesk invitation still open

**Comment by ed-holland** [comment]
Using Chrome on Windows. Clearing cache did not help.

The HM entities are sensors which pick up their value from an entity attribute. No STATE customizing on the.
I state-customizing templates on some other binary sensors to make the message more clear (as on/off). Was not aware that that causes problems.

On the returning colors. More were coming back but not all. It might just be that i did not wait long enough.

**Comment by Mariusthvdb** [comment]
> Using Chrome on Windows. Clearing cache did not help.

using Mac here, so that is beyond my personal options for testing, rely on you for this. did we already check for errors in the logs?

> The HM entities are sensors which pick up their value from an entity attribute. No STATE customizing on the.

That should be fine then. Regular template sensors should update their states correctly, which would then be picked up by the custom_ui. What can happen is, upon reload of template entities, they change state to unknown, for which the template should cater. Always guard your templates. Still reload template entities is not the same as a mere page reload (which in it self isnt the same as a refresh..) and shouldn't cause havoc at all. A mystery.

> I state-customizing templates on some other binary sensors to make the message more clear (as on/off). Was not aware that that causes problems.
> 

Its just that most people dont realize the actual state in the HA state machine changes because of this. Other entities or logic depending on the state of these entities should take that into account. As would customizing templates.

My advise is always to leave the states as they are, and if you need special states, create a template sensor doing so.

> On the returning colors. More were coming back but not all. It might just be that i did not wait long enough.

Maybe. Still, they shouldn't be gone at all. I know Chrome isnt the easiest of browsers clearing cache, so it still might be that. Did you test with another browser? Or with the companion app on a mobile device?

**Comment by ed-holland** [comment]
Yes, we checked errors in inspect and HA log. 
Agree, the offline state is forgotten often also in customizing. Will also add additional sensors for the customizing that has state changes to visualize. Good advise. Thx.

We also checked several browsers.

I think it is an issue from HA itself not triggering the refresh for custom_ui. There have been many changes.

**Comment by Mariusthvdb** [comment]
> We also checked several browsers.

If this is the case, you should change the title of this topic.

But, in all honesty, I don't think it is the true issue. As said, having tested all browsers with various dashboards and configs, custom_ui works as expected, 100%, refresh or not. No matter which device I use, be it an old iPad, iPhone, or the latest state of the art ones, after iOS 10 (which seems to cause issues indeed) all customizations work correctly, provided the custom_ui is installed correctly, and the templates the user throws at it are correct (ie use the true states of the entities, and use a guard for unforeseen states).

> I think it is an issue from HA itself not triggering the refresh for custom_ui. There have been many changes.

Unless you can indicate a true issue in HA (eg by means of a log in Chrome inspector) I think I have shown above this is not the case.
If anything, it is a Chrome thing. But, as you claim, you see it happen in other browsers too? Maybe it is a Windows/Chrome issue, the one aspect I cant test on this side. But even then, if a customization would be able to do its job, not would be noted in a log somewhere.

Will try a windows machine tonight, load Chrome on my fully customized production system, and hope to report back.
Meanwhile, if the title doesn't fully cover the issue, please adapt accordingly, to make it better describe the issue at hand, and make it easier for others to find the correct topic.
Thanks.

**Comment by ed-holland** [comment]
Noticed the same effect as refreshing browser when LoveLace was reloaded (added a new Frontend integration in HACS).

**Comment by dmatora** [comment]
I experience this on safari/chrome/iOS/iPadOS

**Comment by Mariusthvdb** [comment]
So I just experimented on a brand spanking new Windows machine, heavily secured, (causing some extra delay in vpn traffic...) and fired up Chrome. There is not a single customization out of order. Not at startup, not during state changes, and not after hitting the refresh button. Time after time again.

This is a real mystery now, and the only thing left I can think of are templating issues.

**Comment by ed-holland** [comment]
I'm using custom_ui 99% for icon color. The 1% for icons.

We started with the one entity on a dashboard...
![image](https://user-images.githubusercontent.com/53995991/103515233-a5cd5b00-4e6e-11eb-99c8-69592bba516b.png)
Refresh
![image](https://user-images.githubusercontent.com/53995991/103515256-af56c300-4e6e-11eb-9070-4b2d4578484a.png)

HACS brought up a popup window... did not make a screenshot of the exact wording...

**Comment by Mariusthvdb** [comment]
ok, didn't know this was a test view. 

not sure what HACS would have to do with any of this? Can't you make that re-happen?

only thing I can say now is:

> sensor.hm_low_battery:
>   templates:
>     icon_color: if (state == 'O.K.') return 'green'; return 'orange';    
> Any idea why this is not working?

Yes, that won't work, because you either need to quote the template, or use the multi-line notation I use in the examples:

```
sensor.hm_low_battery:
  templates:
    icon_color: >
      if (state == 'O.K.') return 'green';
      return 'orange';
```

Still I think you already adapted that?

**Comment by ed-holland** [comment]
true, that was out of date...  I've changed it to be a counter with the messages as attribute (easier in automation). The messages are shown in lovelace using secondary_info (not in the test dashboard)
```

sensor.hm_low_battery:
  templates:
    icon_color: if (state == '0') return 'green'; return 'red';
```
![image](https://user-images.githubusercontent.com/53995991/103521094-7facb880-4e78-11eb-90fb-da780136b4c6.png)

**Comment by Mariusthvdb** [comment]
yes, but this is the same incorrect template. As said, you should either quote single line templates, but (for me) preferably use multiline notation:

```
sensor.hm_low_battery:
  templates:
    icon_color: >
      if (state == '0') return 'green'; 
      return 'red';
```

**Comment by ed-holland** [comment]
I've changed all customizing accordingly but it makes no difference. The refresh still behaves the same way.
![image](https://user-images.githubusercontent.com/53995991/104088031-2a9be880-5264-11eb-83dc-c4928dada056.png)

I've also found no documentation on YAML standard / HA that describes that single lines require quoting.
[YAML Std](https://en.wikipedia.org/wiki/YAML#Syntax) states that strings do not have to be quoted.
Quoting also needs to be done the right way to avoid issues [HA doc](https://developers.home-assistant.io/docs/documenting/standards#single-vs-double-quotation-marks)

For Readability, just Line Folding would also be an option.

Can you elaborate on why the single line usages is wrong?

**Comment by ed-holland** [comment]
Just captured the HACS Reload
![image](https://user-images.githubusercontent.com/53995991/104088365-9b440480-5266-11eb-93ab-141e64237f12.png)

**Comment by Mariusthvdb** [comment]
> I've also found no documentation on YAML standard / HA that describes that single lines require quoting.
> YAML Std states that strings do not have to be quoted.

I might be mistaken, sorry for that, but please note these are not string, but templates. As long as there is  not a true string after the icon or icon_color field (eg simply red, or mdi:wifi) but a template you should better be safe than sorry. Hence my always using the multi-line notation, where, as you do use, the `return 'mdi:wifi';` is used. If you need that outcome in a larger template, you have to make clear to the interpreter what is a string and what is part of the template.

Using jinja yaml rules has always worked well: https://www.home-assistant.io/integrations/template/#examples

not sure why you post https://developers.home-assistant.io/docs/documenting/standards#single-vs-double-quotation-marks but if you didnt know about that, yes, that's some basic info you need to apply.

Why the screendump of the HACS reload? This has nothing to do with anything custom_ui? 
 
I guess will need to close this issue, because as far as I can find, there's nothing related to custom_ui causing your issue, and nobody else has supported the issue so far.

**Comment by Mariusthvdb** [comment]
CORRECTION...

As above, I stand corrected. After having used this custom_ui for over 4 years now, it is only now I find that the quoting of single line templates isnt necessary after all. Hence, taking out the need (well, need, but at least for me the preferred method) of multi-line templates, in order to prevent exactly that need for quoting.

I've adapted my test setup with this:

```
homeassistant:
  customize_glob:
    '*.*':
      hide_attributes:
        - templates

  customize_domain:
    device_tracker:
      templates:
        icon_color: if (state == 'home') return 'green';return 'purple';

  customize:
    binary_sensor.updater:
      templates:
        icon_color: if (state == 'on') return 'red';return 'green';

    device_tracker.hall:
      templates:
        icon_color: if (state == 'home') return 'green';return 'orange';
        icon: if (state == 'home') return 'mdi:speaker-wireless';return 'mdi:speaker-off';

    person.marius:
      templates:
        entity_picture: return '/local/images/family/marius_bmj_' + state + '.png';
```

and all entities are customized accordingly.

While this still doesnt solve @ed-holland's refresh issue, it is rather noteworthy, and I will add a paragraph to the readme doc and the template examples.

thanks for helping me find out about this!

**Comment by ed-holland** [comment]
> Why the screendump of the HACS reload? This has nothing to do with anything custom_ui?

You wondered why is was using the "reload" instead of "refresh" somewhere in this issue. More a FYI

**Comment by ed-holland** [comment]
Even after a few HA updates, the issue is still there. As it is only a cosmetic issue, it can be closed.
Thanks @Mariusthvdb for the effort. Please continue the good work.

**Comment by dmatora** [comment]
Problem is some entities only show last update time after you do a refresh, so you either see 'red' zones, either when did they become red, which not that much a cosmetic issue.

**Comment by Mariusthvdb** [comment]
well, this is rather unsatisfactory, to say the least. Haven been able to reproduce with custom-ui, in 3 instances. My production system has over 2800 entities, and they are all, repeat all, customized one way or another. None of them show the behavior you describe. Except for a couple of python set entities (which we cant customize with custom-ui, but can set icon_color in the python script itself) and a few other entities, made by custom integrations (like the googledrive backup binary)

However, experimenting with card-mod on a few badges I noticed the same behavior. Maybe you can check and reproduce>

using this:
```
badges:
  - entity: person.marijn
    <<: &badge
      style:
        ha-state-label-badge:
          $:
            ha-label-badge:
              $: |
                .label-badge .label span{
                  {% if is_state(config.entity,'home') %} color: gold;
                  {% else %} color: black;
                  {% endif %}
                }
          .: |
            :host {
              {% set zones = states.zone|map(attribute='name')|list %}
              {% if is_state(config.entity,'home') %} --label-badge-red: green;
              {% elif 'bij' in states(config.entity) %} --label-badge-red: gold;
              {% elif states(config.entity) in ['moving','driving'] or
                      states(config.entity) in zones %} --label-badge-red: mediumslateblue
              {% else %} --label-badge-red: dimgrey;
              {% endif %}
            }
```

makes the styling appear, but only after sometime. If it does, a mere page refresh makes the badges use the default colors, only to have the style kick back in after some tome again. Cant say it has to do with any state change, but this. might be related.

If so, it is not custom related in anyway, but would hint at a HA internal thing.

Got to remember custom-ui is not core, nor anything the HA team endorses at all. So, we;'ve got to be glad it still works as it does (in my setup it is rock solid, even better behaving than the card-mod style).

Ill try report this behavior in the card-mod repo, maybe Thomas has thought about it that could help here too. If so, Ill be back ;-)
